File: app/lib/models/announcement.dart
import 'package:omi/backend/schema/gen/announcements_wire.g.dart' as wire;

enum AnnouncementType { changelog, feature, announcement }

enum TriggerType { immediate, versionUpgrade, firmwareUpgrade }

String _triggerTypeToWire(TriggerType trigger) {
  switch (trigger) {
    case TriggerType.immediate:
      return 'immediate';
    case TriggerType.versionUpgrade:
      return 'version_upgrade';
    case TriggerType.firmwareUpgrade:
      return 'firmware_upgrade';
  }
}

TriggerType _triggerTypeFromWire(String? trigger) {
  switch (trigger) {
    case 'immediate':
      return TriggerType.immediate;
    case 'firmware_upgrade':
      return TriggerType.firmwareUpgrade;
    case 'version_upgrade':
      return TriggerType.versionUpgrade;
    default:
      throw FormatException('Unknown announcement trigger: $trigger');
  }
}

String _announcementTypeToWire(AnnouncementType type) => type.name;

AnnouncementType _announcementTypeFromWire(String type) {
  final announcementType = AnnouncementType.values.asNameMap()[type];
  if (announcementType == null) {
    throw FormatException('Unknown announcement type: $type');
  }
  return announcementType;
}

class Targeting {
  final String? appVersionMin;
  final String? appVersionMax;
  final String? firmwareVersionMin;
  final String? firmwareVersionMax;
  final List<String>? deviceModels;
  final List<String>? platforms;
  final TriggerType trigger;
  final List<String>? testUids;

  Targeting({
    this.appVersionMin,
    this.appVersionMax,
    this.firmwareVersionMin,
    this.firmwareVersionMax,
    this.deviceModels,
    this.platforms,
    this.trigger = TriggerType.versionUpgrade,
    this.testUids,
  });

  factory Targeting.fromJson(Map<String, dynamic> json) {
    return Targeting.fromGenerated(wire.GeneratedTargeting.fromJson(json));
  }

  factory Targeting.fromGenerated(wire.GeneratedTargeting generated) {
    return Targeting(
      appVersionMin: generated.appVersionMin,
      appVersionMax: generated.appVersionMax,
      firmwareVersionMin: generated.firmwareVersionMin,
      firmwareVersionMax: generated.firmwareVersionMax,
      deviceModels: generated.deviceModels ?? const [],
      platforms: generated.platforms ?? const [],
      trigger: _triggerTypeFromWire(generated.trigger),
      testUids: generated.testUids,
    );
  }

  wire.GeneratedTargeting toGenerated() {
    return wire.GeneratedTargeting(
      appVersionMin: appVersionMin,
      appVersionMax: appVersionMax,
      firmwareVersionMin: firmwareVersionMin,
      firmwareVersionMax: firmwareVersionMax,
      deviceModels: deviceModels,
      platforms: platforms,
      trigger: _triggerTypeToWire(trigger),
      testUids: testUids,
    );
  }

  Map<String, dynamic> toJson() => toGenerated().toJson();
}

// Display model for controlling announcement presentation
class Display {
  final int priority;
  final DateTime? startAt;
  final DateTime? expiresAt;
  final bool dismissible;
  final bool showOnce;

  Display({this.priority = 0, this.startAt, this.expiresAt, this.dismissible = true, this.showOnce = true});

  factory Display.fromJson(Map<String, dynamic> json) {
    return Display.fromGenerated(wire.GeneratedDisplay.fromJson(json));
  }

  factory Display.fromGenerated(wire.GeneratedDisplay generated) {
    return Display(
      priority: generated.priority,
      startAt: generated.startAt,
      expiresAt: generated.expiresAt,
      dismissible: generated.dismissible,
      showOnce: generated.showOnce,
    );
  }

  wire.GeneratedDisplay toGenerated() {
    return wire.GeneratedDisplay(
      priority: priority,
      startAt: startAt,
      expiresAt: expiresAt,
      dismissible: dismissible,
      showOnce: showOnce,
    );
  }

  Map<String, dynamic> toJson() => toGenerated().toJson();
}

// Changelog content models
class ChangelogItem {
  final String title;
  final String description;
  final String? icon;

  ChangelogItem({required this.title, required this.description, this.icon});

  factory ChangelogItem.fromJson(Map<String, dynamic> json) {
    return ChangelogItem(
      title: json['title'] as String,
      description: json['description'] as String,
      icon: json['icon'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {'title': title, 'description': description, 'icon': icon};
}

class ChangelogContent {
  final String title;
  final List<ChangelogItem> changes;

  ChangelogContent({required this.title, required this.changes});

  factory ChangelogContent.fromJson(Map<String, dynamic> json) {
    return ChangelogContent(
      title: json['title'] as String,
      changes: (json['changes'] as List<dynamic>)
          .map((item) => ChangelogItem.fromJson(item as Map<String, dynamic>))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() => {'title': title, 'changes': changes.map((item) => item.toJson()).toList()};
}

// Feature content models
class FeatureStep {
  final String title;
  final String description;
  final String? imageUrl;
  final String? videoUrl;
  final String? highlightText;

  FeatureStep({required this.title, required this.description, this.imageUrl, this.videoUrl, this.highlightText});

  factory FeatureStep.fromJson(Map<String, dynamic> json) {
    return FeatureStep(
      title: json['title'] as String,
      description: json['description'] as String,
      imageUrl: json['image_url'] as String?,
      videoUrl: json['video_url'] as String?,
      highlightText: json['highlight_text'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'description': description,
      'image_url': imageUrl,
      'video_url': videoUrl,
      'highlight_text': highlightText,
    };
  }
}

class FeatureContent {
  final String title;
  final List<FeatureStep> steps;

  FeatureContent({required this.title, required this.steps});

  factory FeatureContent.fromJson(Map<String, dynamic> json) {
    return FeatureContent(
      title: json['title'] as String,
      steps:
          (json['steps'] as List<dynamic>).map((item) => FeatureStep.fromJson(item as Map<String, dynamic>)).toList(),
    );
  }

  Map<String, dynamic> toJson() => {'title': title, 'steps': steps.map((step) => step.toJson()).toList()};
}

// Announcement content models
class AnnouncementCTA {
  final String text;
  final String action;

  AnnouncementCTA({required this.text, required this.action});

  factory AnnouncementCTA.fromJson(Map<String, dynamic> json) {
    return AnnouncementCTA(text: json['text'] as String, action: json['action'] as String);
  }

  Map<String, dynamic> toJson() => {'text': text, 'action': action};
}

class AnnouncementContent {
  final String title;
  final String body;
  final String? imageUrl;
  final AnnouncementCTA? cta;

  AnnouncementContent({required this.title, required this.body, this.imageUrl, this.cta});

  factory AnnouncementContent.fromJson(Map<String, dynamic> json) {
    return AnnouncementContent(
      title: json['title'] as String,
      body: json['body'] as String,
      imageUrl: json['image_url'] as String?,
      cta: json['cta'] == null ? null : AnnouncementCTA.fromJson(json['cta'] as Map<String, dynamic>),
    );
  }

  Map<String, dynamic> toJson() => {'title': title, 'body': body, 'image_url': imageUrl, 'cta': cta?.toJson()};
}

// Main announcement model
class Announcement {
  final String id;
  final AnnouncementType type;
  final DateTime createdAt;
  final bool active;

  // Version fields (used by /changelogs endpoint)
  final String? appVersion;
  final String? firmwareVersion;
  final List<String>? deviceModels;
  final DateTime? expiresAt;

  // Flexible targeting and display options
  final Targeting? targeting;
  final Display? display;

  // Raw content - parsed based on type
  final Map<String, dynamic> content;

  Announcement({
    required this.id,
    required this.type,
    required this.createdAt,
    this.active = true,
    this.appVersion,
    this.firmwareVersion,
    this.deviceModels,
    this.expiresAt,
    this.targeting,
    this.display,
    required this.content,
  });

  factory Announcement.fromJson(Map<String, dynamic> json) {
    return Announcement.fromGenerated(wire.GeneratedAnnouncement.fromJson(json));
  }

  factory Announcement.fromGenerated(wire.GeneratedAnnouncement generated) {
    return Announcement(
      id: generated.id,
      type: _announcementTypeFromWire(generated.type),
      createdAt: generated.createdAt,
      active: generated.active,
      appVersion: generated.appVersion,
      firmwareVersion: generated.firmwareVersion,
      deviceModels: generated.deviceModels,
      expiresAt: generated.expiresAt,
      targeting: generated.targeting == null ? null : Targeting.fromGenerated(generated.targeting!),
      display: generated.display == null ? null : Display.fromGenerated(generated.display!),
      content: generated.content,
    );
  }

  wire.GeneratedAnnouncement toGenerated() {
    return wire.GeneratedAnnouncement(
      id: id,
      type: _announcementTypeToWire(type),
      createdAt: createdAt,
      active: active,
      appVersion: appVersion,
      firmwareVersion: firmwareVersion,
      deviceModels: deviceModels,
      expiresAt: expiresAt,
      targeting: targeting?.toGenerated(),
      display: display?.toGenerated(),
      content: content,
    );
  }

  Map<String, dynamic> toJson() => toGenerated().toJson();

  // Type-specific content getters
  ChangelogContent get changelogContent => ChangelogContent.fromJson(content);
  FeatureContent get featureContent => FeatureContent.fromJson(content);
  AnnouncementContent get announcementContent => AnnouncementContent.fromJson(content);

  // Get effective display priority (defaults to 0)
  int get effectivePriority => display?.priority ?? 0;

  // Get effective trigger type
  TriggerType get effectiveTrigger => targeting?.trigger ?? TriggerType.versionUpgrade;
}

File: backend/models/announcement.py
from datetime import datetime
from enum import Enum
from typing import Any, List, Mapping, Optional, cast

from pydantic import BaseModel


class AnnouncementType(str, Enum):
    CHANGELOG = "changelog"
    FEATURE = "feature"
    ANNOUNCEMENT = "announcement"


class TriggerType(str, Enum):
    IMMEDIATE = "immediate"  # Check every app launch
    VERSION_UPGRADE = "version_upgrade"  # Check only when app version changes
    FIRMWARE_UPGRADE = "firmware_upgrade"  # Check only when firmware version changes


class Targeting(BaseModel):
    """Controls who sees the announcement"""

    app_version_min: Optional[str] = None  # Show to users >= this version
    app_version_max: Optional[str] = None  # Show to users <= this version
    firmware_version_min: Optional[str] = None
    firmware_version_max: Optional[str] = None
    device_models: Optional[List[str]] = None  # ["Omi DevKit 2", "Omi Pro"]
    platforms: Optional[List[str]] = None  # ["ios", "android"]
    trigger: TriggerType = TriggerType.VERSION_UPGRADE
    test_uids: Optional[List[str]] = None  # If set, only these users see the announcement (for testing)

    def to_dict(self) -> dict[str, Any]:
        return {
            "app_version_min": self.app_version_min,
            "app_version_max": self.app_version_max,
            "firmware_version_min": self.firmware_version_min,
            "firmware_version_max": self.firmware_version_max,
            "device_models": self.device_models,
            "platforms": self.platforms,
            "trigger": self.trigger.value,
            "test_uids": self.test_uids,
        }


class Display(BaseModel):
    """Controls how/when the announcement is displayed"""

    priority: int = 0  # Higher = show first
    start_at: Optional[datetime] = None  # Don't show before this time
    expires_at: Optional[datetime] = None  # Don't show after this time
    dismissible: bool = True  # Can user skip?
    show_once: bool = True  # Only show once per user

    def to_dict(self) -> dict[str, Any]:
        return {
            "priority": self.priority,
            "start_at": self.start_at,
            "expires_at": self.expires_at,
            "dismissible": self.dismissible,
            "show_once": self.show_once,
        }


def _nested_dict(value: Any) -> Optional[dict[str, Any]]:
    if not isinstance(value, Mapping):
        return None
    return dict(cast(Mapping[str, Any], value))


# Changelog content models
class ChangelogItem(BaseModel):
    title: str
    description: str
    icon: Optional[str] = None


class ChangelogContent(BaseModel):
    title: str
    changes: List[ChangelogItem]


# Feature content models
class FeatureStep(BaseModel):
    title: str
    description: str
    image_url: Optional[str] = None
    video_url: Optional[str] = None
    highlight_text: Optional[str] = None


class FeatureContent(BaseModel):
    title: str
    steps: List[FeatureStep]


# Announcement content models
class AnnouncementCTA(BaseModel):
    text: str
    action: str  # e.g., "navigate:/settings/premium" or "url:https://example.com"


class AnnouncementContent(BaseModel):
    title: str
    body: str
    image_url: Optional[str] = None
    cta: Optional[AnnouncementCTA] = None


# Main announcement model
class Announcement(BaseModel):
    id: str
    type: AnnouncementType
    created_at: datetime
    active: bool = True

    # Legacy version triggers (for backward compatibility with existing announcements)
    app_version: Optional[str] = None
    firmware_version: Optional[str] = None
    device_models: Optional[List[str]] = None

    # Legacy expiration (for backward compatibility)
    expires_at: Optional[datetime] = None

    # New flexible targeting and display options (optional)
    targeting: Optional[Targeting] = None
    display: Optional[Display] = None

    # Content - will be one of ChangelogContent, FeatureContent, or AnnouncementContent
    content: dict[str, Any]

    def get_changelog_content(self) -> ChangelogContent:
        return ChangelogContent(**self.content)

    def get_feature_content(self) -> FeatureContent:
        return FeatureContent(**self.content)

    def get_announcement_content(self) -> AnnouncementContent:
        return AnnouncementContent(**self.content)

    def get_effective_targeting(self) -> Targeting:
        """Get targeting config, falling back to legacy fields if not set."""
        if self.targeting:
            return self.targeting
        # Build targeting from legacy fields
        return Targeting(
            app_version_min=self.app_version,
            app_version_max=self.app_version,
            firmware_version_min=self.firmware_version,
            firmware_version_max=self.firmware_version,
            device_models=self.device_models,
            trigger=TriggerType.VERSION_UPGRADE,
        )

    def get_effective_display(self) -> Display:
        """Get display config, falling back to legacy fields if not set."""
        if self.display:
            return self.display
        # Build display from legacy fields
        return Display(
            expires_at=self.expires_at,
        )

    @staticmethod
    def from_dict(data: Mapping[str, Any]) -> "Announcement":
        targeting_data = _nested_dict(data.get("targeting"))
        display_data = _nested_dict(data.get("display"))

        # Tolerate a missing or out-of-enum type so one malformed/legacy announcement document cannot
        # 500 the whole (public) announcements list. Fall back to the generic ANNOUNCEMENT type.
        raw_type = data.get("type")
        try:
            announcement_type = AnnouncementType(raw_type)
        except ValueError:
            announcement_type = AnnouncementType.ANNOUNCEMENT

        return Announcement(
            id=cast(str, data.get("id")),
            type=announcement_type,
            created_at=cast(datetime, data.get("created_at")),
            active=data.get("active", True),
            app_version=data.get("app_version"),
            firmware_version=data.get("firmware_version"),
            device_models=data.get("device_models"),
            expires_at=data.get("expires_at"),
            targeting=Targeting(**targeting_data) if targeting_data else None,
            display=Display(**display_data) if display_data else None,
            content=data.get("content", {}),
        )

    def to_dict(self) -> dict[str, Any]:
        result = {
            "id": self.id,
            "type": self.type.value,
            "created_at": self.created_at,
            "active": self.active,
            "app_version": self.app_version,
            "firmware_version": self.firmware_version,
            "device_models": self.device_models,
            "expires_at": self.expires_at,
            "content": self.content,
        }
        if self.targeting:
            result["targeting"] = self.targeting.to_dict()
        if self.display:
            result["display"] = self.display.to_dict()
        return result


# API Response models
class ChangelogResponse(BaseModel):
    changelogs: List[Announcement]


class FeatureResponse(BaseModel):
    features: List[Announcement]


class AnnouncementListResponse(BaseModel):
    announcements: List[Announcement]
